### PR TITLE
Add PushLimitThroughOuterJoin rule

### PR DESCRIPTION
**Pull request by kasiafi**
Ref https://github.com/prestodb/presto/pull/11971

### CODE REVIEW

**Comment by martint** [comment]
@cla-bot check

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.RIGHT;
+import static io.prestosql.sql.planner.plan.Patterns.Join.type;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+
+public class PushLimitThroughOuterJoin
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: Could you add a Javadoc how the rule changes plans, see: `io.prestosql.sql.planner.iterative.rule.TransformExistsApplyToLateralNode`
for example

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Override
+    public Result apply(LimitNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        if (isLeftOuter(joinNode.getType()) && !isLimited(left, context.getLookup(), parent.getCount())) {
+            left = new LimitNode(context.getIdAllocator().getNextId(), left, parent.getCount(), false);
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: make it partial limit as it relaxes partitioning constraints, see: `io.prestosql.sql.planner.optimizations.AddLocalExchanges.Rewriter#visitLimit`

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        if (isRightOuter(joinNode.getType()) && !isLimited(right, context.getLookup(), parent.getCount())) {
+            right = new LimitNode(context.getIdAllocator().getNextId(), right, parent.getCount(), false);
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: ditto

**Comment by kasiafi** [comment]
Comments applied.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return Result.empty();
+    }
+
+    private static boolean isLimited(PlanNode node, Lookup lookup, long limit)
+    {
+        Range<Long> cardinality = extractCardinality(node, lookup);
+        return cardinality.hasUpperBound() && cardinality.upperEndpoint() <= limit;
+    }
+
+    private static boolean isLeftOuter(Type type)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: isLeftOrFullOuter

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        return type == LEFT || type == FULL;
+    }
+
+    private static boolean isRightOuter(Type type)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: isRightOrFullOuter

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushLimitThroughOuterJoin.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                                    p.values(5, leftKey),
+                                    p.values(5, rightKey),
+                                    new EquiJoinClause(leftKey, rightKey)));
+                })
+                .matches(
+                        limit(1,
+                                join(
+                                        FULL,
+                                        ImmutableList.of(equiJoinClause("leftKey", "rightKey")),
+                                        limit(1, values("leftKey")),
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushLimitThroughOuterJoin.java]: Could you validate new limit is partial?

**Comment by kasiafi** [comment]
Comments applied.

**Comment by kokosing** [comment]
Merged, thanks!

**Pull request merged**